lib/galaxy/model/security.py

```diff
@@ -1447,62 +1439,171 @@ def get_showable_folders(
-            for group in groups:
-                self.associate_components(group=group, role=role)
+        Associations are set only if a list of new associations is provided.
+        If the provided list is empty, existing associations will be removed.
+        If the provided value is None, existing associations will not be updated.
+        """
+        self._ensure_model_instance_has_id(group)
+        if user_ids is not None:
+            self._set_group_users(group, user_ids)
+        if role_ids is not None:
+            self._set_group_roles(group, role_ids)
+        # Commit only if both group users and group roles have been set.
+        self.sa_session.commit()
+
+    def set_role_user_and_group_associations(
+        self,
+        role: Role,
+        *,
+        user_ids: Optional[List[int]] = None,
+        group_ids: Optional[List[int]] = None,
+    ) -> None:
+        """
+        Set role users and role groups, replacing current associations.
+
+        Associations are set only if a list of new associations is provided.
+        If the provided list is empty, existing associations will be removed.
+        If the provided value is None, existing associations will not be updated.
+        """
+        self._ensure_model_instance_has_id(role)
+        if user_ids is not None:
+            self._set_role_users(role, user_ids or [])
+        if group_ids is not None:
+            self._set_role_groups(role, group_ids or [])
+        # Commit only if both role users and role groups have been set.
+        self.sa_session.commit()
+
+    def _set_user_groups(self, user, group_ids):
+        delete_stmt = delete(UserGroupAssociation).where(UserGroupAssociation.user_id == user.id)
+        insert_values = [{"user_id": user.id, "group_id": group_id} for group_id in group_ids]
+        self._set_associations(user, UserGroupAssociation, delete_stmt, insert_values)
+
+    def _set_user_roles(self, user, role_ids):
+        # Do not include user's private role association in delete statement.
+        delete_stmt = delete(UserRoleAssociation).where(UserRoleAssociation.user_id == user.id)
+        private_role = get_private_user_role(user, self.sa_session)
+        if not private_role:
+            log.warning("User %s does not have a private role assigned", user)
+        else:
+            delete_stmt = delete_stmt.where(UserRoleAssociation.role_id != private_role.id)
+        role_ids = self._filter_private_roles(role_ids)
+        insert_values = [{"user_id": user.id, "role_id": role_id} for role_id in role_ids]
+        self._set_associations(user, UserRoleAssociation, delete_stmt, insert_values)
+
+    def _filter_private_roles(self, role_ids):
+        """Filter out IDs of private roles: those should not be assignable via UI"""
+        filtered = []
+        for role_id in role_ids:
+            stmt = select(Role.id).where(Role.id == role_id).where(Role.type == Role.types.PRIVATE)
```

Comment:
I think you can reduce this to a single select statement with something like `Role.id.in_(role_ids)`

Reply:
Thanks!